README.md
## UNDER DEVELOPMENT

# Blender Power Sequencer Proxy

This is a [standalone python package](https://pypi.org/project/bpsproxy/) as well as a module which is used under the hood in the [Blender Power Sequencer add-on](https://github.com/GDquest/Blender-power-sequencer) to generate better proxies to be used in Blender.


## Install

It can be installed as a standalone command line utility [via PiPy](https://pypi.org/project/bpsproxy/): `pip install [--user] bpsproxy`. *Note* that you have to have `$HOME/.local/bin` included in your `$PATH` environment variable (on unix) if you're going to install the utility locally (using `--user` when executing `pip`).


## Usage

After installing the script, get help by writing `bpsproxy -h`:

```
usage: bpsproxy [-h] [-p {webm,mp4,nvenc}] [-s {25,50,100} [{25,50,100} ...]]
                [-v] [--dry-run]
                [working_directory]

Create proxies for Blender VSE using FFMPEG.

positional arguments:
  working_directory     The directory containing media to create proxies for

optional arguments:
  -h, --help            show this help message and exit
  -p {webm,mp4,nvenc}, --preset {webm,mp4,nvenc}
                        a preset name for proxy encoding
  -s {25,50,100} [{25,50,100} ...], --sizes {25,50,100} [{25,50,100} ...]
                        A list of sizes of the proxies to render, either 25,
                        50, or 100
  -v, --verbose         Increase verbosity level (eg. -vvv).
  --dry-run             Run the script without actual rendering or creating
                        files and folders. For DEBUGGING purposes
```

## External Dependencies

`BPSProxy` requires

- `ffmpeg`
- `ffprobe`

to be available in the `$PATH` environment variable in order to work. In case `BPSProxy` will catch a missing dependency it will throw a message error similar to this:

```
ERROR:BPS:BPSProxy couldn't find external dependencies:
[X] ffmpeg: NOT FOUND
[X] ffprobe: NOT FOUND
Check if you have them properly installed and available in the PATH environemnt variable.
~ Exiting.
```


CHANGELOG.md
# Blender Proxies Generator: Changelog

This document lists new features, improvements, changes, and bug fixes brought
to the project.

## BPSProxy 0.2.1

### Bug fixes

- Preserve the rotation of the input video so that the proxies have the same
  orientation as the input in Blender. #7




bpsproxy/call.py
# import multiprocessing as mp
import os
import subprocess as sp
import sys

from functools import partial
from itertools import chain, tee
from tqdm import tqdm
from .config import LOGGER
from .utils import get_dir, kickstart

WINDOWS = ("win32", "cygwin")


def call_makedirs(cfg, clargs, **kwargs):
    """
    Make BL_proxy directories if necessary.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    kwargs: dict
    MANDATORY: path_i
    Dictionary with additional information from previous step.
    """
    path_i = kwargs["path_i"]
    path_d = map(partial(get_dir, cfg, clargs, **kwargs), path_i)
    path_d = tee(chain(*path_d))
    kickstart(map(lambda p: LOGGER.info("Directory @ {}".format(p)), path_d[0]))
    if clargs.dry_run:
        return
    path_d = (os.makedirs(p, exist_ok=True) for p in path_d[1])
    kickstart(path_d)


def call(cfg, clargs, *, cmds, **kwargs):
    """
    Generic subprocess calls.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    kwargs: dict
    MANDATORY: path_i
    Dictionary with additional information from previous step.

    Returns
    -------
    out: str
    Stdout & Stderr gathered from subprocess call.
    """
    kwargs_s = {
        "stdout": sp.PIPE,
        "stderr": sp.STDOUT,
        "universal_newlines": True,
        "check": kwargs.get("check", True),
        "shell": kwargs.get("shell", False),
        "creationflags": sp.CREATE_NEW_PROCESS_GROUP if sys.platform in WINDOWS else 0,
    }
    if kwargs_s["shell"]:
        cmds = map(lambda cmd: (cmd[0], " ".join(cmd[1])), cmds)
    cmds = tee(cmds)
    kickstart(map(lambda cmd: LOGGER.debug("CALL :: {}".format(cmd[1])), cmds[0]))
    if clargs.dry_run:
        return []
    n = len(kwargs["path_i"])
    ps = tqdm(
        map(lambda cmd: sp.run(cmd[1], **kwargs_s), cmds[1]),
        total=n,
        unit="file" if n == 1 else "files",
    )
    return [p.stdout for p in ps]


bpsproxy/utils.py
"""
Collection of utility functions, class-independent
"""
import os.path as osp
from collections import deque
from shutil import which


class ToolError(Exception):
    """Raised if external dependencies aren't found on system.
    """

    pass


def checktools(tools):
    tools = [(t, which(t) or "") for t in tools]
    check = {"tools": tools, "test": all(map(lambda x: x[1], tools))}
    if not check["test"]:
        msg = ["BPSProxy couldn't find external dependencies:"]
        msg += [
            "[{check}] {tool}: {path}".format(
                check="v" if path is not "" else "X", tool=tool, path=path or "NOT FOUND"
            )
            for tool, path in check["tools"]
        ]
        msg += [
            (
                "Check if you have them properly installed and available in the PATH"
                " environemnt variable."
            )
        ]
        raise ToolError("\n".join(msg))


def get_path_video(cfg, clargs, path, **kwargs):
    return osp.join(
        osp.dirname(path), cfg["proxy_directory"], osp.basename(path), "proxy_{size}.avi"
    )


def get_path_image(cfg, clargs, path, **kwargs):
    return osp.join(
        osp.dirname(path),
        cfg["proxy_directory"],
        "images",
        "{size}",
        "{file}_proxy.jpg".format(file=osp.basename(path)),
    )


def get_path(cfg, clargs, path, **kwargs):
    get_path_f = {"video": get_path_video, "image": get_path_image}
    what = what_vi(cfg, clargs, path, **kwargs)
    return get_path_f[what](cfg, clargs, path, **kwargs)


def get_dir_video(cfg, clargs, path, **kwargs):
    return iter((osp.join(osp.dirname(path), cfg["proxy_directory"], osp.basename(path)),))


def get_dir_image(cfg, clargs, path, **kwargs):
    ps = osp.join(osp.dirname(path), cfg["proxy_directory"], "images", "{size}")
    return map(lambda s: ps.format(size=s), clargs.sizes)


def get_dir(cfg, clargs, path, **kwargs):
    get_dir_f = {"video": get_dir_video, "image": get_dir_image}
    what = what_vi(cfg, clargs, path, **kwargs)
    return get_dir_f[what](cfg, clargs, path, **kwargs)


def what_vi(cfg, clargs, p, **kwargs):
    return "video" if osp.splitext(p)[1].lower() in cfg["extensions"]["video"] else "image"


def kickstart(it):
    deque(it, maxlen=0)


def printw(cfg, text, s="\n", e="...", p="", **kwargs):
    p = p or cfg["pre"]["work"]
    print("{s}{p} {}{e}".format(text, s=s, e=e, p=p), **kwargs)


def printd(cfg, text, s="", e=".", p="", **kwargs):
    p = p or cfg["pre"]["done"]
    printw(cfg, text, s=s, e=e, p=p, **kwargs)


def prints(cfg, text, s="", e=".", p="", **kwargs):
    p = p or cfg["pre"]["skip"]
    printw(cfg, text, s=s, e=e, p=p, **kwargs)


bpsproxy/commands.py
import os.path as osp
import shlex as sl
from itertools import chain
from .utils import get_path


def get_commands_check(cfg, clargs, **kwargs):
    """
    ffprobe subprocess command generation.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    kwargs: dict
    MANDATORY: path_i_1, path_o_1
    Dictionary with additional information from previous step.

    Returns
    -------
    out: iter(tuple(str))
    Iterator containing commands.
    """
    cmd = (
        "ffprobe -v error -select_streams v:0 -show_entries stream=nb_frames -of"
        " default=noprint_wrappers=1:nokey=1 '{file}'"
    )
    out = map(lambda s: kwargs["path_o_1"].format(size=s), clargs.sizes)
    out = map(lambda f: cmd.format(file=f), out)
    out = sl.split(cmd.format(file=kwargs["path_i_1"]) + " && " + " && ".join(out))
    return iter((out,))


def get_commands_image_1(cfg, clargs, **kwargs):
    """
    ffmpeg subprocess command generation for processing an image.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    kwargs: dict
    MANDATORY: path_i_1, path_o_1
    Dictionary with additional information from previous step.

    Returns
    -------
    out: iter(tuple(str))
    Iterator containing commands.
    """
    cmd = "ffmpeg -y -v quiet -stats -i '{path_i_1}' {common_all}"
    common = "-f apng -filter:v scale=iw*{size}:ih*{size} '{path_o_1}'"
    common_all = map(lambda s: kwargs["path_o_1"].format(size=s), clargs.sizes)
    common_all = map(
        lambda s: common.format(size=s[0] / 100.0, path_o_1=s[1]), zip(clargs.sizes, common_all)
    )
    common_all = " ".join(common_all)
    out = sl.split(cmd.format(path_i_1=kwargs["path_i_1"], common_all=common_all))
    return iter((out,))


def get_commands_video_1(cfg, clargs, **kwargs):
    """
    ffmpeg subprocess command generation for processing a video.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    kwargs: dict
    MANDATORY: path_i_1, path_o_1
    Dictionary with additional information from previous step.

    Returns
    -------
    out: iter(tuple(str))
    Iterator containing commands.
    """
    cmd = "ffmpeg -y -v quiet -stats -i '{path_i_1}' {common_all}"
    common = (
        "-pix_fmt yuv420p"
        " -g 1"
        " -sn -an"
        " -vf colormatrix=bt601:bt709"
        " -vf scale=ceil(iw*{size}/2)*2:ceil(ih*{size}/2)*2"
        " {preset}"
        " '{path_o_1}'"
    )
    common_all = map(lambda s: kwargs["path_o_1"].format(size=s), clargs.sizes)
    common_all = map(
        lambda s: common.format(
            preset=cfg["presets"][clargs.preset], size=s[0] / 100.0, path_o_1=s[1]
        ),
        zip(clargs.sizes, common_all),
    )
    common_all = " ".join(common_all)
    out = sl.split(cmd.format(path_i_1=kwargs["path_i_1"], common_all=common_all))
    return iter((out,))


def get_commands(cfg, clargs, *, what, **kwargs):
    """
    Delegates the creation of commands lists to appropriate functions based on `what` parameter.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    what: str
    Determines the returned value (see: Returns[out]).
    kwargs: dict
    MANDATORY: path_i
    Dictionary with additional information from previous step.

    Returns
    -------
    out: iter(tuple(str, tuple(str)))
    An iterator with the 1st element as a tag (the `what` parameter) and the 2nd
    element as the iterator of the actual commands.
    """
    get_commands_f = {
        "video": get_commands_video_1,
        "image": get_commands_image_1,
        "check": get_commands_check,
    }
    ps = (
        kwargs["path_i"]
        if what not in cfg["extensions"]
        else filter(
            lambda p: osp.splitext(p)[1].lower() in cfg["extensions"][what], kwargs["path_i"]
        )
    )
    ps = map(lambda p: (p, get_path(cfg, clargs, p, **kwargs)), ps)
    out = chain.from_iterable(
        map(lambda p: get_commands_f[what](cfg, clargs, path_i_1=p[0], path_o_1=p[1], **kwargs), ps)
    )
    return map(lambda c: (what, c), out)


def get_commands_vi(cfg, clargs, **kwargs):
    """
    Delegates the creation of commands lists to appropriate functions for video/image processing.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.
    clargs: Namespace
    Command line arguments.
    cmds: iter(tuple(str))
    kwargs: dict
    MANDATORY: path_i_1, path_o_1
    Dictionary with additional information from previous step.

    Returns
    -------
    out: iter(tuple(str, tuple(str)))
    An iterator with the 1st element as a tag (the `what` parameter) and the 2nd
    element as the iterator of the actual commands.
    """
    ws = filter(lambda x: x is not "all", cfg["extensions"])
    return chain.from_iterable(map(lambda w: get_commands(cfg, clargs, what=w, **kwargs), ws))


bpsproxy/__main__.py
"""
Tool to render video proxies using FFMPEG
Offers mp4 and webm options
"""
import argparse as ap
import glob as g
import logging as lg
import os.path as osp
import sys
from itertools import compress, starmap, tee

from .call import call, call_makedirs
from .commands import get_commands, get_commands_vi
from .config import CONFIG as C
from .config import LOGGER, LOGLEV
from .utils import checktools, printw, printd, prints, ToolError


def find_files(
    directory=".", ignored_directory=C["proxy_directory"], extensions=C["extensions"]["all"]
):
    """
    Find files to process.

    Parameters
    ----------
    directory: str
    Working directory.
    ignored_directory: str
    Don't check for files in this directory. By default `BL_proxy`.
    extensions: set(str)
    Set of file extensions for filtering the directory tree.

    Returns
    -------
    out: list(str)
    List of file paths to be processed.
    """
    if not osp.isdir(directory):
        raise ValueError(("The given path '{}' is not a valid directory.".format(directory)))
    xs = g.iglob("{}/**".format(osp.abspath(directory)), recursive=True)
    xs = filter(lambda x: osp.isfile(x), xs)
    xs = filter(lambda x: ignored_directory not in osp.dirname(x), xs)
    xs = [x for x in xs if osp.splitext(x)[1].lower() in extensions]
    return xs


def parse_arguments(cfg):
    """
    Uses `argparse` to parse the command line arguments.

    Parameters
    ----------
    cfg: dict
    Configuration dictionary.

    Returns
    -------
    out: Namespace
    Command line arguments.
    """
    p = ap.ArgumentParser(description="Create proxies for Blender VSE using FFMPEG.")
    p.add_argument(
        "working_directory",
        nargs="?",
        default=".",
        help="The directory containing media to create proxies for",
    )
    p.add_argument(
        "-p",
        "--preset",
        default="mp4",
        choices=cfg["presets"],
        help="a preset name for proxy encoding",
    )
    p.add_argument(
        "-s",
        "--sizes",
        nargs="+",
        type=int,
        default=[25],
        choices=cfg["proxy_sizes"],
        help="A list of sizes of the proxies to render, either 25, 50, or 100",
    )
    p.add_argument(
        "-v", "--verbose", action="count", default=0, help="Increase verbosity level (eg. -vvv)."
    )
    p.add_argument(
        "--dry-run",
        action="store_true",
        help=(
            "Run the script without actual rendering or creating files and"
            " folders. For DEBUGGING purposes"
        ),
    )

    clargs = p.parse_args()
    # normalize directory
    clargs.working_directory = osp.abspath(clargs.working_directory)
    # --dry-run implies maximum verbosity level
    clargs.verbose = 99999 if clargs.dry_run else clargs.verbose
    return clargs


def main():
    """
    Script entry point.
    """
    tools = ["ffmpeg", "ffprobe"]
    try:
        # get command line arguments and set log level
        clargs = parse_arguments(C)
        lg.basicConfig(level=LOGLEV[min(clargs.verbose, len(LOGLEV) - 1)])

        # log basic command line arguments
        clargs.dry_run and LOGGER.info("DRY-RUN")
        LOGGER.info("WORKING-DIRECTORY :: {}".format(clargs.working_directory))
        LOGGER.info("PRESET :: {}".format(clargs.preset))
        LOGGER.info("SIZES :: {}".format(clargs.sizes))

        # check for external dependencies
        checktools(tools)

        # find files to process
        path_i = find_files(clargs.working_directory)
        kwargs = {"path_i": path_i}

        printw(C, "Creating directories if necessary")
        call_makedirs(C, clargs, **kwargs)

        printw(C, "Checking for existing proxies")
        cmds = tee(get_commands(C, clargs, what="check", **kwargs))
        stdouts = call(C, clargs, cmds=cmds[0], check=False, shell=True, **kwargs)
        checks = map(lambda s: s.strip().split(), stdouts)
        checks = starmap(lambda fst, *tail: not all(fst == t for t in tail), checks)
        kwargs["path_i"] = list(compress(kwargs["path_i"], checks))

        if len(kwargs["path_i"]) != 0:
            printw(C, "Processing", s="\n")
            cmds = get_commands_vi(C, clargs, **kwargs)
            call(C, clargs, cmds=cmds, **kwargs)
        else:
            printd(C, "All proxies exist or no files found, nothing to process", s="\n")
        printd(C, "Done")
    except (ToolError, ValueError) as e:
        LOGGER.error(e)
        prints(C, "Exiting")
    except KeyboardInterrupt:
        prints(C, "DirtyInterrupt. Exiting", s="\n\n")
        sys.exit()


# this is so it can be ran as a module: `python3 -m bpsrender` (for testing)
if __name__ == "__main__":
    main()
